packages/tpetra/core/src/Tpetra_Details_lclMultiply.hpp

```diff
@@ -0,0 +1,149 @@
+            const XMV& B_lcl,
+            typename XMV::value_type beta,
+            XMV& C_lcl)
+{
+  using Teuchos::CONJ_TRANS;
+  using Teuchos::NO_TRANS;
+  using Teuchos::TRANS;
+  using Kokkos::ALL;
+  using Kokkos::subview;
+  using VT = typename XMV::value_type;
+  using ATS = Kokkos::ArithTraits<VT>;
+
+
+  const char ctransA = (transA == NO_TRANS ? 'N' :
+                        (transA == TRANS ? 'T' : 'C'));
+  const char ctransB = (transB == NO_TRANS ? 'N' :
+                        (transB == TRANS ? 'T' : 'C'));
+
+  const size_t zero = 0;
+  const VT zeroVT = ATS::zero();
+  const VT oneVT = ATS::one();
+
+
+  // The first two cases are specializations for inner products between a vector and a multivector
+  if (beta == zeroVT && transA != NO_TRANS && transB == NO_TRANS && A_lcl.extent(1) == 1) {
+    auto C_sub = subview (C_lcl, zero, ALL ());
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-dot");
+    KokkosBlas::dot(C_sub, A_lcl, B_lcl);
+    if (alpha != oneVT)
+      KokkosBlas::scal (C_sub, alpha, C_sub);
+  }
+  else if (beta == zeroVT && transA != NO_TRANS && transB == NO_TRANS && B_lcl.extent(1) == 1) {
+    auto C_sub = subview (C_lcl, ALL (), zero );
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-dot");
+    KokkosBlas::dot(C_sub, A_lcl, B_lcl);
+    if (alpha != oneVT)
+      KokkosBlas::scal (C_sub, alpha, C_sub);
+  }
+  else {
+    auto C_sub = subview (C_lcl,
+                          std::make_pair (zero, C_lcl.extent(0)),
+                          std::make_pair (zero, C_lcl.extent(1)));
+    ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-gemm");
+    KokkosBlas::gemm(&ctransA, &ctransB, alpha, A_lcl, B_lcl, beta, C_sub);
+  }
+}
+
+template<class XMV,
+         typename std::enable_if<!std::is_same<typename XMV::value_type,typename Kokkos::Details::InnerProductSpaceTraits<typename XMV::value_type>::dot_type>::value>::type* = nullptr >
+void
+lclMultiply(const Teuchos::ETransp transA,
+            const Teuchos::ETransp transB,
+            typename XMV::value_type alpha,
+            const XMV& A_lcl,
+            const XMV& B_lcl,
+            typename XMV::value_type beta,
+            XMV& C_lcl)
+{
+  using Teuchos::CONJ_TRANS;
+  using Teuchos::NO_TRANS;
+  using Teuchos::TRANS;
+  using Kokkos::ALL;
+  using Kokkos::subview;
+
+
+  const char ctransA = (transA == NO_TRANS ? 'N' :
+                        (transA == TRANS ? 'T' : 'C'));
+  const char ctransB = (transB == NO_TRANS ? 'N' :
+                        (transB == TRANS ? 'T' : 'C'));
+
+  const size_t zero = 0;
+
+  auto C_sub = subview (C_lcl,
+                        std::make_pair (zero, C_lcl.extent(0)),
+                        std::make_pair (zero, C_lcl.extent(1)));
+  ::Tpetra::Details::ProfilingRegion regionGemm ("Tpetra::MV::multiply-call-gemm");
+  KokkosBlas::gemm(&ctransA, &ctransB, alpha, A_lcl, B_lcl, beta, C_sub);
```

Comment:
1. This is the case where `dot_type != value_type`.  How is it possible for `C_lcl` to have the same
type as `A_lcl` and `B_lcl`?

2. `XMV::value_type` could be either const or nonconst.  It's fine to use for `alpha` and `beta`, where constness doesn't matter, but just remember that if you want the nonconst type, you'll need to use `XMV::non_const_value_type`.

Reply:
`MV::multiply` multiplies two MVs into `this`. So `C_lcl` is the memory of a MV, not `dot_type`. Is that what you are getting at?

Reply:
Ah, never mind, sorry.